### PR TITLE
ReleaseTrain Role Based Access Control

### PR DESCRIPTION
**Pull request by bernardo-bastos**
Adds Role Based Access Control to `ReleaseTrain`.

Rev: DSN-J1WO8X

### CODE REVIEW

File: services/cd-service/pkg/repository/transformer.go [hunk 1]
```diff
@@ -1050,6 +1051,11 @@ func (c *ReleaseTrain) Transform(ctx context.Context, state *State) (string, err
 		if envConfig.Upstream == nil {
 			return fmt.Sprintf("Environment %q does not have upstream configured - exiting.", envName), nil
 		}
+		err := state.checkUserPermissions(ctx, envName, "*", auth.PermissionDeployReleaseTrain, c.RBACConfig)
+		if err != nil {
+			return "", err
+		}
+
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer.go:1058]: here (down below) when calling the deploy transformer, we need to pass along the authentication.

This actually needs to be done, in all cases when a transformer calls another transformer

File: services/cd-service/pkg/repository/transformer_test.go [hunk 2]
```diff
@@ -793,6 +793,119 @@ func TestRbacTransformerTest(t *testing.T) {
...
+						Upstream: &config.EnvironmentConfigUpstream{
+							Environment: envAcceptance, // train drives from acceptance to production
+						},
+					},
+				},
+				&CreateEnvironment{
+					Environment: envAcceptance,
+					Config: config.EnvironmentConfig{
+						Upstream: &config.EnvironmentConfigUpstream{
+							Environment: envAcceptance,
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer_test.go:811]: don't use same env

File: services/cd-service/pkg/repository/transformer_test.go [hunk 3]
```diff
@@ -793,6 +793,119 @@ func TestRbacTransformerTest(t *testing.T) {
...
+				},
+				&DeployApplicationVersion{
+					Environment: envAcceptance,
+					Application: "test",
+					Version:     2,
+				},
+				&ReleaseTrain{
+					Target: envProduction,
+					Authentication: Authentication{RBACConfig: auth.RBACConfig{DexEnabled: true, Policy: map[string]*auth.Permission{
+						"developer,DeployReleaseTrain,production:production,*,allow": {Role: "developer"},
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer_test.go:848]: we should also need deploy permission to do this

File: services/cd-service/pkg/repository/transformer_test.go [hunk 4]
```diff
@@ -793,6 +793,119 @@ func TestRbacTransformerTest(t *testing.T) {
...
+				},
+				&CreateApplicationVersion{
+					Application: "test",
+					Manifests: map[string]string{
+						envProduction: "productionmanifest",
+						envAcceptance: "acceptancenmanifest",
+					},
+				},
+				&DeployApplicationVersion{
+					Environment: envProduction,
```
Comment (sven-urbanski-freiheit-com) [inline services/cd-service/pkg/repository/transformer_test.go:824]: lots of the same transformers -> extract in a function or have a seperate test array for these 2 cases

**Review by sven-urbanski-freiheit-com — CHANGES_REQUESTED** [review]
.

File: services/cd-service/pkg/repository/transformer.go (hunk 1 above)
Comment (bernardo-bastos) [reply services/cd-service/pkg/repository/transformer.go:1058]: Added the authentication check for `CreateApplicationVersion` as well. Added a test for it.

File: services/cd-service/pkg/repository/transformer_test.go (hunk 2 above)
Comment (bernardo-bastos) [reply services/cd-service/pkg/repository/transformer_test.go]: Fixed the test issue.

File: services/cd-service/pkg/repository/transformer_test.go (hunk 3 above)
Comment (bernardo-bastos) [reply services/cd-service/pkg/repository/transformer_test.go]: Added the authentication check for CreateApplicationVersion as well. Added a test for it.

**Pull request merged**